---
output:
  md_document:
    variant: markdown_github
---

```{r, image, echo = FALSE, fig.align='center', out.width=300, out.height=300}
knitr::include_graphics("/Users/carlo.manenti/OneDrive - Htechnopole/carlo/RESEARCH_PROJECTS/PACKAGES/davilalab/scOps/scOps.png")
```


## A package for single-cell operational representations.

\n
\n

### scOps installation for Python.

scOps comes in both Python and R. Please select the language (and so the repository branch) that best suites your need, since the implementation is virtually identical. In both cases we use standard objects to store the representations, *AnnData* or a *SingleCellExperiment* object, respectively.

To install scOps for R please use `pip install git+https:<.Python_version1.0.0.>` specifying the URL of Python_version1.0.0. Or download the Python_version1.0.0 folder and use `pip install .` after moving inside that folder. 

### scOps representations: BOGs, Signatures and Profiles.

scOps offers three main types of representations: Bag Of Genes, Profiles and Signatures.

-   ***Bag Of Genes*** **(BOGs)** can be obtained via  `scOps.compute.BOGs()` (*Python*); They are the differentially expressed genes of a cell type/condition, with respect to all the others.

-   ***Signatures*** can be obtained via `scOps.compute.Signatures()` (*Python*); As the name suggests, this representation highlights the most distinctive features of each cell type/condition. Higher (lower) scores correspond to preferentially expressed (down regulated) features.

-   ***Profiles*** can be obtained via `scOps.compute.Profiles()` (*Python*); They represent the average transcriptome of a given cell type/condition by averaging together the counts for each gene.

### scOps maintenance.

To this day scOps is maintained by Carlo and Erik. Please open an issue for any doubt, question, or bug you find using the package. Any feedback is always welcomed.
